[Excel](custom functions) update tutorial to latest correct steps (#2945)
OfficeDev/office-js-docs-pr: docs/tutorials/share-data-and-events-between-custom-functions-and-the-task-pane-tutorial.md

---
title: "Tutorial: Share data and events between Excel custom functions and the task pane"
description: 'Learn how to share data and events between custom functions and the task pane in Excel.'
ms.date: 09/17/2021
ms.prod: excel
ms.localizationpriority: high
---

# Tutorial: Share data and events between Excel custom functions and the task pane

You can configure your Excel add-in to use a shared runtime. This makes it possible to shared global data, or send events between the task pane and custom functions. For most custom functions scenarios, we recommend using a shared runtime, unless you have a specific reason to use a non-task pane (UI-less) custom function. This tutorial assumes you're familiar with using the Yo Office generator to create add-in projects. Consider completing the [Excel custom functions tutorial](excel-tutorial-create-custom-functions.md), if you haven't already.

## Create the add-in project

You'll use the [Yeoman generator for Office Add-ins](https://github.com/OfficeDev/generator-office) to create the Excel add-in project.

> - To generate an Excel add-in with custom functions, run the command `yo office --projectType excel-functions --name 'Excel shared runtime add-in' --host excel --js true`.

Reviewer: Can we put this yo office command in the command line code box so it's easily copied?

Author: good idea!

> The generator will create the project and install supporting Node components.

Reviewer: Small tense suggestion.
```suggestion
The generator creates the project and installs supporting Node components.
```

## Configure the manifest

Follow these steps to configure the add-in project to use a shared runtime.

1. Start Visual Studio Code and open the add-in project you generated.
1. Open the **manifest.xml** file.

> 1. Update the requirements section to use the [shared runtime](../reference/requirement-sets/shared-runtime-requirement-sets.md) instead of the custom function runtime. The XML should appear as follows.

Reviewer: Should we be explicit that "requirements" is an XML element?
```suggestion
1. Update the `<Requirements>` section to use the [shared runtime](../reference/requirement-sets/shared-runtime-requirement-sets.md) instead of the custom function runtime. The XML should appear as follows.
```

    ```xml
    <Hosts>
      <Host Name="Workbook"/>
    </Hosts>
    <Requirements>
      <Sets DefaultMinVersion="1.1">
        <Set Name="SharedRuntime" MinVersion="1.1"/>
      </Sets>
    </Requirements>
    <DefaultSettings>

>     ```

Reviewer: I like having the context for where Requirements is in the file, but I don't like having to copy everything when only the one part is getting replaced.

Author: It's tricky because over time we've had issues where people get confused about where to put some of the XML, or accidentally put it inside another tag and such. So I've added some context about where it is located to try to help. But yeah, then you can't just copy paste...

1. Find the `<VersionOverrides>` section and add the following `<Runtimes>` section. The lifetime needs to be **long** so that your add-in code can run even when the task pane is closed. The `resid` value is **Taskpane.Url**, which references the **taskpane.html** file location specified in the ` <bt:Urls>` section near the bottom of the **manifest.xml** file.

    > [!IMPORTANT]
    > The `<Runtimes>` section must be entered after the `<Host>` element in the exact order shown in the following XML.

   ```xml
   <VersionOverrides ...>
     <Hosts>
       <Host ...>
         <Runtimes>
           <Runtime resid="Taskpane.Url" lifetime="long" />
         </Runtimes>
       ...

>        </Host>

Reviewer: It's a little confusing that this `<Host>` element doesn't match the one in the previous step.

Author: Well, unfortunately, that is how the manifest is designed. This is actually a different section. Again why I try to include more details to help know where you are making updates.

Reviewer: ... I may have messed up this manifest, which would explain the error. Let me try again, but better!

   ```

1. Find the `<Page>` element. Then change the source location from **Functions.Page.Url** to **Taskpane.Url**.

   ```xml
   <AllFormFactors>
   ...
   <Page>
     <SourceLocation resid="Taskpane.Url"/>
   </Page>
   ...
   ```

1. Find the `<FunctionFile ...>` tag and change the `resid` from **Commands.Url** to  **Taskpane.Url**.

    ```xml
    </GetStarted>
    ...
    <FunctionFile resid="Taskpane.Url"/>
    ...
    ```

1. Save the **manifest.xml** file.

## Configure the webpack.config.js file

The **webpack.config.js** will build multiple runtime loaders. You need to modify it to load only the shared JavaScript runtime via the **taskpane.html** file.

> 1. Start Visual Studio Code and open the Excel or PowerPoint add-in project you generated.

Reviewer: You can remove this step since we've already opened VS Code.

1. Open the **webpack.config.js** file.
1. Remove the following **functions.html** plugin code.

    ```javascript
    new HtmlWebpackPlugin({
        filename: "functions.html",
        template: "./src/functions/functions.html",
        chunks: ["polyfill", "functions"]
      })
    ```

1. Remove the following **commands.html** plugin code.

    ```javascript
    new HtmlWebpackPlugin({
        filename: "commands.html",
        template: "./src/commands/commands.html",
        chunks: ["polyfill", "commands"]
      })
    ```

1. Add **functions** and **commands** entries to the chunks list as shown next.

    ```javascript
      new HtmlWebpackPlugin({
        filename: "taskpane.html",
        template: "./src/taskpane/taskpane.html",
        chunks: ["polyfill", "taskpane", "commands", "functions"]
      })

>     ```

Reviewer: Consider describing the changes, then providing the final `plugins` code bock. That would make it quicker to copy and paste the correct code.

1. Save your changes and rebuild the project.

   ```command line
   npm run build
   ```

> [!NOTE]

> > You can also remove the **functions.html** and **commands.html** files. The **taskpane.html** will load the **functions.js** and **commands.js** code into the shared JavaScript runtime via the webpack updates you just made.

Reviewer: This note isn't indented under the previous step, which is also causing the next step to appear as "1".

![image](https://user-images.githubusercontent.com/38896772/133852399-5e934972-4013-4c4a-8b75-e9387ddd6b18.png)

1. Save your changes and run the project. Ensure that it loads and runs with no errors.

   ```command line

>    npm start

Reviewer: I got this error in the workbook that opens for this step:
![image](https://user-images.githubusercontent.com/38896772/133852490-e21fa28b-ca51-4a17-8017-311233800925.png)

Reviewer: Okay, fixing the manifest fixed this error. Shocking, I know...

1. Find the `<Page>` element. Then change the source location from **Functions.Page.Url** to **Taskpane.Url**.

   ```xml
   <AllFormFactors>
   ...
   <Page>
     <SourceLocation resid="Taskpane.Url"/>
   </Page>
   ...
   ```

1. Find the `<FunctionFile ...>` tag and change the `resid` from **Commands.Url** to  **Taskpane.Url**.

    ```xml
    </GetStarted>
    ...
    <FunctionFile resid="Taskpane.Url"/>
    ...
    ```

1. Save the **manifest.xml** file.

## Configure the webpack.config.js file

The **webpack.config.js** will build multiple runtime loaders. You need to modify it to load only the shared JavaScript runtime via the **taskpane.html** file.

1. Open the **webpack.config.js** file.
1. Go to the `plugins:` section.

> 1. Remove the following `functions.html` plugin if it esists

Reviewer: ```suggestion
1. Remove the following `functions.html` plugin if it exists.
```

    
    ```javascript
    new HtmlWebpackPlugin({
        filename: "functions.html",
        template: "./src/functions/functions.html",
        chunks: ["polyfill", "functions"]
      })
    ```

1. Remove the following `commands.html` plugin if it exists.

    ```javascript
    new HtmlWebpackPlugin({
        filename: "commands.html",
        template: "./src/commands/commands.html",
        chunks: ["polyfill", "commands"]
      })
    ```

1. If you removed the `functions` or `commands` plugin, add them as `chunks`. The following JavaScript shows the updated entry if you removed both `functions` and `commands` plugins.
    
    ```javascript
      new HtmlWebpackPlugin({
        filename: "taskpane.html",
        template: "./src/taskpane/taskpane.html",
        chunks: ["polyfill", "taskpane", "commands", "functions"]
      })
    ```
    
1. Save your changes and rebuild the project.

>    ```command line

Reviewer: ```suggestion
   ```command&nbsp;line
```

   npm run build
   ```
    
    > [!NOTE]

>     > You can also remove the **functions.html** and **commands.html** files. The **taskpane.html** will load the **functions.js** and **commands.js** code into the shared JavaScript runtime via the webpack updates you just made.

Reviewer: ```suggestion
    > You can also remove the **functions.html** and **commands.html** files. The **taskpane.html** loads the **functions.js** and **commands.js** code into the shared JavaScript runtime via the webpack updates you just made.
```

    
1. Save your changes and run the project. Ensure that it loads and runs with no errors.
    

>    ```command line

Reviewer: ```suggestion
   ```command&nbsp;line
```

> 

Reviewer: Small nit, but in the last step we're asked to call `npm run start`. Two things:
- We called `npm start` earlier. Should this be the same command?
- The tutorial never has the user close the add-in. Should we have them close and reopen? Or should we remove the earlier start? FWIW, I never closed the previous run and didn't get the new functions. Closing and rerunning npm start refreshed things.
